WebDriver BiDi: support cookie-related commands (storage domain)

AnujPanta1 opened the pull request:
#### b1f15f6fc24a45d8a148299aff2c3cc95b47bda5
<pre>
WebDriver BiDi: support cookie-related commands (storage domain)
<a href="https://rdar.apple.com/136802953">rdar://136802953</a>

Reviewed by BJ Burg.

Implementing more core functionality within the storage module. Introducing the ability
to get, set and delete cookies by accessing cookie stores. Currently storage keys are
not supported, so browsingContext is the main method of identifying cookie stores.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp: Added.
(WebKit::BidiStorageAgent::BidiStorageAgent):
(WebKit::buildObjectForCookie):
(WebKit::buildArrayForCookies):
(WebKit::matchesCookieFilter):
(WebKit::BidiStorageAgent::makePartitionKey):
(WebKit::BidiStorageAgent::cookieStoreForPartition):
(WebKit::BidiStorageAgent::getCookies):
(WebKit::BidiStorageAgent::setCookie):
(WebKit::deleteCookiesSequentially):
(WebKit::BidiStorageAgent::deleteCookies):
* Source/WebKit/UIProcess/Automation/BidiStorageAgent.h: Added.
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::WebDriverBidiProcessor):
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
* Source/WebKit/UIProcess/Automation/protocol/BidiStorage.json: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/295759@main">https://commits.webkit.org/295759@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f8b508b2e8583348d068cd612d546dd37528fc93

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104480 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24190 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14607 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109687 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55151 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24683 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32734 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79331 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107486 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19195 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94296 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59658 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18977 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54520 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88674 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12426 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112072 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31641 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23305 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88372 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32005 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90529 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88038 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22766 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32933 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10708 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/26118 "The change is no longer eligible for processing.") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31568 "Build is in progress. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Running apply-patch; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36908 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31360 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34699 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32920 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/WebKit/CMakeLists.txt:
```diff
@@ -954,6 +954,7 @@ set(WebKit_WEBDRIVER_BIDI_PROTOCOL_GENERATOR_INPUTS
     ${WEBKIT_DIR}/UIProcess/Automation/protocol/BidiBrowsingContext.json
     ${WEBKIT_DIR}/UIProcess/Automation/protocol/BidiLog.json
     ${WEBKIT_DIR}/UIProcess/Automation/protocol/BidiScript.json
+    ${WEBKIT_DIR}/UIProcess/Automation/protocol/BidiStorage.json
```
Nit: alphabetize

burg commented on Source/WebKit/DerivedSources.make:
```diff
@@ -519,6 +519,7 @@ WEBDRIVER_BIDI_PROTOCOL_INPUT_FILES = \
     $(WebKit2)/UIProcess/Automation/protocol/BidiBrowsingContext.json \
     $(WebKit2)/UIProcess/Automation/protocol/BidiLog.json \
     $(WebKit2)/UIProcess/Automation/protocol/BidiScript.json \
+	$(WebKit2)/UIProcess/Automation/protocol/BidiStorage.json \
```
Nit: alphabetize, and indentation seems off.

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright (C) 2025 Apple Inc. All rights reserved.
+ * Copyright (C) 2025 Microsoft Corporation. All rights reserved.
```
Style nit: this second line isn't needed until someone from Microsoft commits on the file, or the file had been copied from a file with a Microsoft line.

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,292 @@
+#include "BidiStorageAgent.h"
+
+#if ENABLE(WEBDRIVER_BIDI)
+
```
Nit: extra newline

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,292 @@
+#if ENABLE(WEBDRIVER_BIDI)
+
+
+#include "AutomationProtocolObjects.h"
```
What is this include needed for?

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,292 @@
+    , m_storageDomainDispatcher(BidiStorageBackendDispatcher::create(backendDispatcher, this))
+{
+}
+BidiStorageAgent::~BidiStorageAgent() = default;
```
Nit: needs newline

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,292 @@
+        .setDomain(cookie.domain)
+        .release();
+
+    partialCookie->setPath(cookie.path);
```
Style nit: unless setting the fields depends on some condition, I think these later `setField()` calls could be chained with the required fields. If it's possible to express it this way, we can avoid declaring a local variable entirely.

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiStorage.json:
```diff
@@ -0,0 +1,140 @@
+        {
+            "id": "PartitionKey",
+            "type": "object",
+            "description": "The storage.PartitionKey type represents a storage partition key",
```
Nit: use \<code\>\</code\> around code in cross references in the spec files.

burg replied:
Nit: other cross references have been using the 'domain' name rather than the 'exposedAs' name. So, BidiStorage.PartitionKey.

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiStorage.json:
```diff
@@ -0,0 +1,140 @@
+{
+    "domain": "BidiStorage",
+    "exposedAs": "storage",
+    "description": "The storage module contains functionality and events related to storage",
```
Style nit: in a comment, a period is expected at the end of a sentence. (This is WebKit style.)

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiStorage.json:
```diff
@@ -0,0 +1,140 @@
+            "spec": "https://w3c.github.io/webdriver-bidi/#type-storage-PartitionKey",
+            "properties": [
+                { "name": "userContext", "type": "string", "optional": true, "description": "A user context id" },
+                { "name": "sourceOrigin", "type": "string", "optional": true, "description": "The serialization of the orgin of resources that can access the storage partition" }
```
Nit: `.` and orgin -> origin

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiStorage.json:
```diff
@@ -0,0 +1,140 @@
+            "description": "The storage.PartitionKey type represents a storage partition key",
+            "spec": "https://w3c.github.io/webdriver-bidi/#type-storage-PartitionKey",
+            "properties": [
+                { "name": "userContext", "type": "string", "optional": true, "description": "A user context id" },
```
This should be a cross reference to `BidiBrowser.UserContext`, which is a string.

burg replied:
Nit: `.` at the end. The comment for this property seems optional once we use the proper type as suggested above.

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiStorage.json:
```diff
@@ -0,0 +1,140 @@
+                { "name": "size", "type": "integer", "optional": true, "description": "The size of the cookie" },
+                { "name": "httpOnly", "type": "boolean", "optional": true, "description": "If the cookie is HTTP only" },
+                { "name": "secure", "type": "boolean", "optional": true, "description": "If the cookie is secure" },
+                { "name": "sameSite", "type": "string", "optional": true, "description": "The SameSite attribute of the cookie" },
```
Nit: use \<code\>

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiStorage.json:
```diff
@@ -0,0 +1,140 @@
+            "description": "Describes a storage partition based on a browsing context",
+            "spec": "https://w3c.github.io/webdriver-bidi/#type-storage-BrowsingContextPartitionDescriptor",
+            "properties": [
+                { "name": "type", "type": "string", "const": "context", "description": "The type of partition descriptor" },
```
I don't think `const` is a thing?

The way that this construction is expressed in a few other places is to have an enum type that defines the possible values. Take a look in `BidiScript.json` the `EvaluateResulType`, `LocalValueType`, `RemoteValueType` for examples.

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiStorage.json:
```diff
@@ -0,0 +1,140 @@
+            "description": "Describes a storage partition based on a storage key",
+            "spec": "https://w3c.github.io/webdriver-bidi/#type-storage-StorageKeyPartitionDescriptor",
+            "properties": [
+                { "name": "type", "type": "string", "const": "storageKey", "description": "The type of partition descriptor" },
```
Ditto above, this should reference `BidiStorage.PartitionDescriptorType`

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiStorage.json:
```diff
@@ -0,0 +1,140 @@
+            "id": "PartitionDescriptor",
+            "type": "object",
+            "description": "Describes a storage partition",
+            "spec": "https://w3c.github.io/webdriver-bidi/#type-storage-PartitionDescriptor",
```
This is a dead link. I would link to: https://w3c.github.io/webdriver-bidi/#command-storage-getCookies

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiStorage.json:
```diff
@@ -0,0 +1,140 @@
+            "description": "Describes a storage partition",
+            "spec": "https://w3c.github.io/webdriver-bidi/#type-storage-PartitionDescriptor",
+            "properties": [
+                { "name": "browsingContextPartitionDescriptor", "$ref": "BrowsingContextPartitionDescriptor", "optional": true  },
```
I don't think this will work, the spec does not define a top level `browsingContextPartitionDescriptor` key. I think this needs to be flattened, so:


`BidiStorage.PartitionDescriptor`
- type : `BidiStorage.PartitionDescriptorType`
- context: optional `BidiBrowsingContext.BrowsingContext`
- userContext: optional `BidiBrowser.UserContext`
- sourceOrigin: optional String

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiStorage.json:
```diff
@@ -0,0 +1,140 @@
+            "description": "Enumerates values for cookies same site policy",
+            "spec": "https://w3c.github.io/webdriver-bidi/#type-storage-CookieSameSitePolicy",
+            "enum": [
+                "None",
```
Nit: no need to make this split across lines.

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiStorage.json:
```diff
@@ -0,0 +1,140 @@
+            "id": "CookieSameSitePolicy",
+            "type": "string",
+            "description": "Enumerates values for cookies same site policy",
+            "spec": "https://w3c.github.io/webdriver-bidi/#type-storage-CookieSameSitePolicy",
```
This link doesn't work. Please double check your links, if you can't find an anchor for the type definition, then link to its first use instead (getCookie or setCookie?)

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiStorage.json:
```diff
@@ -0,0 +1,140 @@
+                { "name": "path", "type": "string", "optional": true, "description": "The path of the cookie" },
+                { "name": "httpOnly", "type": "boolean", "optional": true, "description": "Whether the cookie is HTTP only" },
+                { "name": "secure", "type": "boolean", "optional": true, "description": "Whether the cookie is secure" },
+                { "name": "sameSite", "$ref": "CookieSameSitePolicy", "optional": true, "description": "The SameSite attribute of the cookie" },
```
Nit: use \<code\> for SameSite

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiStorage.json:
```diff
@@ -0,0 +1,140 @@
+            "spec": "https://w3c.github.io/webdriver-bidi/#command-storage-getCookies",
+            "wpt": "https://github.com/web-platform-tests/wpt/tree/master/webdriver/tests/bidi/storage/get_cookies",
+            "parameters": [
+                { "name": "filter", "$ref": "CookieFilter", "optional": true, "description": "Filter parameters for cookie retrieval" },
```
Please always use the full qualification, i.e., `BidiStorage.CookieFilter` in case the code gets moved around.

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiStorage.json:
```diff
@@ -0,0 +1,140 @@
+            "wpt": "https://github.com/web-platform-tests/wpt/tree/master/webdriver/tests/bidi/storage/delete_cookies",
+            "parameters": [
+                { "name": "filter", "$ref": "CookieFilter", "optional": true, "description": "Filter parameters for cookie deletion" },
+                { "name": "partition", "$ref": "PartitionDescriptor", "optional": true, "description": "The storage partition to delete cookies from" }
```
Style nit: rewrite these comments to not end in "from" or "in". For example:

The storage partition in which to delete cookies.

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.h:
```diff
@@ -0,0 +1,57 @@
+    BidiStorageAgent(WebAutomationSession&, Inspector::BackendDispatcher&);
+    ~BidiStorageAgent() override;
+
+    Inspector::Protocol::ErrorStringOr<Ref<Inspector::Protocol::BidiStorage::PartitionKey>> makePartitionKey(RefPtr<JSON::Object> partitionDescriptor);
```
Nit: does not need to be in the header unless used outside this class.

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,292 @@
+    String value = cookie->getString("value"_s);
+    String domain = cookie->getString("domain"_s);
+
+    if (name.isEmpty() || value.isEmpty() || domain.isEmpty()) {
```
As written, this will hang because it doesn't call `completionHandler`. I think it can be put inside `ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF` so that it returns correctly.

burg replied:
Still needs to be addressed.

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,292 @@
+        }
+
+        auto builtCookies = buildArrayForCookies(matchingCookies);
+
```
Nit: `builtCookies` could be inlined.

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,292 @@
+    RefPtr session = m_session.get();
+    ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!session, InternalError);
+
+    auto keyOrErr = makePartitionKey(optionalPartition);
```
I would call this `parsedPartitionKey`... by convention parsing is always fallible so the null check will make sense as written.

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,292 @@
+    return nullptr;
+}
+
+void BidiStorageAgent::getCookies(RefPtr<JSON::Object>&& optionalFilter, RefPtr<JSON::Object>&& optionalPartition, Inspector::CommandCallbackOf<Ref<JSON::ArrayOf<Inspector::Protocol::BidiStorage::PartialCookie>>, Ref<Inspector::Protocol::BidiStorage::PartitionKey>>&& callback)
```
Style nit: I recommend that you do `using PartitionKey = ...` and similar to shorten signatures in the .cpp files.

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,292 @@
+
+RefPtr<API::HTTPCookieStore> BidiStorageAgent::cookieStoreForPartition(RefPtr<JSON::Object> partitionDescriptor)
+{
+    if (!partitionDescriptor) {
```
I am pretty sure that we should just error out in this case right? it has to be one of `context`, `userContext`, or `sourceOrigin`.

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,292 @@
+    }
+
+    auto type = partitionDescriptor->getString("type"_s);
+    if (type == "context"_s) {
```
We should parse "type" into the enum from BidiProtocolObjects and then use switch cases to handle these. If "type" doesn't parse to a known value then fail.

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,292 @@
+    return { WTFMove(partitionKey) };
+}
+
+RefPtr<API::HTTPCookieStore> BidiStorageAgent::cookieStoreForPartition(RefPtr<JSON::Object> partitionDescriptor)
```
I think this could use our normal sync result type and that would allow returning "Invalid Argument" in these error cases. Right?

i.e., `CommandResult<Ref<API::HTTPCookieStore>>`

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,292 @@
+{
+    if (!partitionDescriptor) {
+        auto page = m_session->webPageProxyForHandle("default"_s);
+        return page ? &page->protectedWebsiteDataStore()->cookieStore() : nullptr;
```
Seems weird to return RefPtr from Ref here, since we have CommandResult

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,292 @@
+        matchingCookies.reserveInitialCapacity(cookiesList.size());
+
+        for (const auto& cookie : cookiesList) {
+            if (matchesCookieFilter(cookie, filter))
```
Naming nit: `cookieMatchesFilter`

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,292 @@
+    ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!keyOrErr, InternalError);
+    auto partitionKey = keyOrErr.value();
+
+    auto cookieStore = cookieStoreForPartition(optionalPartition);
```
It might be preferable to construct the Filter object here so that if it's bogus we don't have to fetch cookies at all.

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,292 @@
+    if (!optionalFilter)
+        return true;
+
+    String filterNameOptional = optionalFilter->getString("name"_s);
```
Naming: `optional` is always a prefix, so `optionalFilterName` and so on.

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,292 @@
+Inspector::Protocol::ErrorStringOr<Ref<Inspector::Protocol::BidiStorage::PartitionKey>> BidiStorageAgent::makePartitionKey(RefPtr<JSON::Object> partitionDescriptor)
+{
+    auto partitionKey = Inspector::Protocol::BidiStorage::PartitionKey::create().release();
+    if (!partitionDescriptor)
```
I believe this should return Invalid Argument if there isn't any partition descriptor?

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,292 @@
+            partitionKey->setSourceOrigin(pageURL.string());
+
+    } else if (type == "storageKey"_s) {
+        // FIXME: support storage key
```
Please add a bugzilla link.

burg reviewed (changes requested):
Overall this is in pretty good shape

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiStorage.json:
```diff
@@ -0,0 +1,145 @@
+            "wpt": "https://github.com/web-platform-tests/wpt/tree/master/webdriver/tests/bidi/storage/set_cookie",
+            "parameters": [
+                { "name": "cookie", "$ref": "PartialCookie", "description": "The cookie to set." },
+                { "name": "partition", "$ref": "PartitionDescriptor", "optional": true, "description": "The storage partition in which to set cookies" }
```
Nit: trailing period

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiStorage.json:
```diff
@@ -0,0 +1,145 @@
+            "wpt": "https://github.com/web-platform-tests/wpt/tree/master/webdriver/tests/bidi/storage/get_cookies",
+            "parameters": [
+                { "name": "filter", "$ref": "BidiStorage.CookieFilter", "optional": true, "description": "Filter parameters for cookie retrieval." },
+                { "name": "partition", "$ref": "PartitionDescriptor", "optional": true, "description": "The storage partition in which to get cookies." }
```
(applies throughout) Please use `BidiStorage` prefix to fully qualify all type references. I see it's done for CookieFilter here but not PartitionDescriptor or others.

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,302 @@
+            callback({ partitionKey });
+            return;
+        }
+        LOG(Automation, "deleteCookies: %zu cookies matched; deleting one-by-one.", toDelete.size());
```
Nit: add newline before LOG

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,302 @@
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(InvalidParameter);
+
+    cookieStore->cookies(
+        [ callback = WTFMove(callback)
```
I know it's kinda ugly, but please match the existing style for lambda capture lists (no newlines/aligned commas). Feel free to align them as necessary during debugging/development for readability, but take them out before landing.

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,302 @@
+
+        Vector<WebCore::Cookie> toDelete;
+        toDelete.reserveInitialCapacity(all.size());
+        for (auto& c : all) {
```
Naming: `for (auto& cookie : fetchedCookies) {`

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,302 @@
+        [ callback = WTFMove(callback)
+        , filter   = WTFMove(optionalFilter)
+        , partitionKey = WTFMove(partitionKey)
+        , cookieStore ](Vector<WebCore::Cookie>&& all) mutable {
```
Naming: `fetchedCookies`

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiStorage.json:
```diff
@@ -0,0 +1,145 @@
+            "type": "string",
+            "description": "Enumerates values for cookies same site policy.",
+            "spec": "https://w3c.github.io/webdriver-bidi/#command-storage-getCookies",
+            "enum": [ "None", "Lax", "Strict"
```
Trailing square bracket should not be on its own line here, it's part of "enum"

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,302 @@
+static void deleteCookiesSequentially(API::HTTPCookieStore* store, Vector<WebCore::Cookie> cookies, size_t index, Ref<PartitionKey> partitionKey, Inspector::CommandCallback<Ref<PartitionKey>> callback)
+{
+    if (index >= cookies.size()) {
+        LOG(Automation, "deleteCookies all deletions complete.");
```
Since there's only one way to get to this code, I don't think this logging is adding anything. We can infer that the deletions are done if the `deleteCookies` command had returned.

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,302 @@
+        auto page = m_session->webPageProxyForHandle(browsingContextHandle);
+        if (!page)
+            SYNC_FAIL_WITH_PREDEFINED_ERROR(InternalError);
+        return { page->protectedWebsiteDataStore()->cookieStore() };
```
Nit: add newline

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,302 @@
+    if (!optionalFilter)
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(InvalidParameter);
+
+
```
Nit: delete extra newline

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,302 @@
+
+        callback({ { buildArrayForCookies(matchingCookies), partitionKey } });
+    });
+
```
Nit: extra newline

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,302 @@
+
+void BidiStorageAgent::deleteCookies(RefPtr<JSON::Object>&& optionalFilter, RefPtr<JSON::Object>&& optionalPartition, Inspector::CommandCallback<Ref<PartitionKey>>&& callback)
+{
+
```
Nit: extra newline

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,302 @@
+    ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!parsedCookieStore, InternalError);
+    Ref<API::HTTPCookieStore> cookieStore = parsedCookieStore.value();
+
+    if (!optionalFilter)
```
I don't think this is correct. If the 'filter' argument is not supplied in the command, then `optionalFilter` will be nullptr and we should not apply filters. But this code as written would return an error in that case.

burg commented on Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp:
```diff
@@ -0,0 +1,302 @@
+    ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!session, InternalError);
+
+    auto parsedPartitionKey = makePartitionKey(optionalPartition);
+    ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!parsedPartitionKey, InternalError);
```
Won't this error out if a filter is passed but not a partition key?

Based on my reading of the spec for this endpoint, filters are applied after obtaining the cookie store from the partition key (using a default partition key if not provided). If there are no filters it should still work.

AnujPanta1 committed:
WebDriver BiDi: support cookie-related commands (storage domain)
rdar://136802953

Reviewed by BJ Burg.

Implementing more core functionality within the storage module. Introducing the ability
to get, set and delete cookies by accessing cookie stores. Currently storage keys are
not supported, so browsingContext is the main method of identifying cookie stores.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp: Added.
(WebKit::BidiStorageAgent::BidiStorageAgent):
(WebKit::buildObjectForCookie):
(WebKit::buildArrayForCookies):
(WebKit::matchesCookieFilter):
(WebKit::BidiStorageAgent::makePartitionKey):
(WebKit::BidiStorageAgent::cookieStoreForPartition):
(WebKit::BidiStorageAgent::getCookies):
(WebKit::BidiStorageAgent::setCookie):
(WebKit::deleteCookiesSequentially):
(WebKit::BidiStorageAgent::deleteCookies):
* Source/WebKit/UIProcess/Automation/BidiStorageAgent.h: Added.
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::WebDriverBidiProcessor):
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
* Source/WebKit/UIProcess/Automation/protocol/BidiStorage.json: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/295759@main
